### DISCUSSION TITLE
Switching to OpenTelemetry from built-in metrics and tracing?

### DISCUSSION THREAD

**Discussion by icebob** [moleculerjs/moleculer#1125] [Ideas]
As I see the trends in tracing and metrics topics, I see the [OpenTelemetry](https://opentelemetry.io/) is becoming the de-facto industry standard for tracing and metrics. OpenTelemetry has mostly all features that have the Moleculer built-in tracing & metrics modules, additionally, it has prebuilt instrumentation modules for many common libraries and is supported by many cloud APM services. 

The biggest disadvantage of the built-in tracing is that it's [hard to connect](https://github.com/moleculerjs/moleculer/discussions/1062) the Moleculer spans to other external lib spans. OpenTelemetry can solve it if you use it everywhere in your project. 

Since both have the same functionality, it's not worth developing the built-in tracing & metrics modules while there is an out-of-box solution with a bigger time & budget capacity background.
_While it would be a breaking change, we can switch it only in the next major version._

What do you think about it? Give a 👍 if you support the switching and give 👎 if not and we should keep and maintain the Moleculer built-in modules or share your thoughts in a comment.

Thanks!

**Comment by shawnmcknight** [comment]
@kthompson23 tagging you on this as I think you'd probably have some opinions here.

**Comment by intech** [comment]
This contribution will provide more benefits in the long term than problems now due to breaking changes.
So I fully support this development. :+1:

**Reply by icebob** [reply]
Thanks.

**Comment by ccampanale** [comment]
I personally support the decision to switch to and adopt OpenTelemetry. I think the writing is on the wall with this one. The "standard" seems to be increasingly adopted widely in the industry which I believe will continue. Adopting would likely better position the framework and limit the amount of custom metrics and tracing development necessity moving forward.

I implemented the AWS X-ray SDK in a custom tracing exporter for the system my team maintains and while it wasn't too complicated, there is a lot of room for improvement. That said, I've placed continued efforts on the back-burner due to possibilities of incorporating OpenTelemetry. I would imagine that incorporating OpenTelemetry would mean easily switching between different providers which would be ideal for teams like mine. 

I think it's the right move. 👍🏻

**Reply by icebob** [reply]
Thanks.

**Comment by adam-mccormick** [comment]
I'm all for this. There's ongoing interest in OpenTelemerty at work and it would be great to see this supported by Moleculer.

**Comment by 1xtr** [comment]
Is it possible to trace methods in services with out create spans?

**Reply by intech** [reply]
@1xtr Tracing is a graph of spans. The more spans, the greater the performance degradation. Therefore, any part of the code that you want to cover the trace will need to be wrapped. If you want to wrap all methods, you can use the [hook `localMethod` in middleware](https://moleculer.services/docs/0.14/middlewares.html#localMethod-next-method).

**Reply by intech** [reply]
@1xtr https://gist.github.com/intech/503facbc5320056d8f327070557fdf31

**Comment by Aaron-Ritter** [comment]
Definitely an upvote from me on OpenTelemetry. But key input for me is first get the most up to date out of the box instrumentation for NodeJS and related frameworks used, and only then try to customize what ever is missing.

OpenTelemetry has a massive backing from the community as well as all the major APM and Cloud vendors on the market.

**Comment by kthompson23** [comment]
This may be out of scope for the first pass and is separate from the OpenTelemetry tracing specification but the OpenTelemetry collector supports gRPC out of the box. With smaller payloads and compression options it could be a real performance benefit.

**Comment by lucwj** [comment]
I'm looking forward OTel features. 👍  💯

**Comment by lucwj** [comment]
Do we have any update here?

**Reply by icebob** [reply]
It won't be implemented in 0.15 yet, instead, I will add an OpenTelemetry-based middleware and you can change the built-in tracing middleware to this OTEL-based one.

**Reply by lucwj** [reply]
> It won't be implemented in 0.15 yet, instead, I will add an OpenTelemetry-based middleware and you can change the built-in tracing middleware to this OTEL-based one.

Don't you mind to share that middleware once you completed?

**Reply by icebob** [reply]
It is in the next branch: https://github.com/moleculerjs/moleculer/tree/next/examples/opentelemetry

**Reply by goggi** [reply]
Hi @icebob, any example on how to integrate the channels to the opentelemetry middleware?

**Comment by Carlvebbesen** [comment]
Is there any update on this issue? and when it will be a part of the molecular js framework?

**Reply by icebob** [reply]
Check this comment: https://github.com/moleculerjs/moleculer/discussions/1125#discussioncomment-6683203
